### PR TITLE
chore: add modal to UI kit

### PR DESCRIPTION
**Pull request by trentwatt**
## Description

WEB-916

Adds modal to the UI kit. [Figma is here](https://www.figma.com/file/Pqs6TVjjHQohXSwJ9qaqwH/Determined-UI-Kit?node-id=261%3A8511&t=Jq08iikMSjHmXimx-1). Particular emphasis was placed on creating developer-friendly modal interface, both for creators of specific modals, and for consumers of those modals, as well as upon preventing the types of longstanding modal state issues we've had with the existing approach.

## Test Plan

No release testing required

## Technical Considerations (edited)
For consumers of specific modals, the API looks like this
```tsx
const ProjectModal = useModalComponent(ProjectModalComponent) 

const props: ProjectModalComponentProps = {...}

return (
    <div>
        <Button onClick={ProjectModal.open} />
        <ProjectModal.Component {...props} />
    </div>
)
```

For Modal creators, the requirement is just to create a component for the body of the modal, that calls `useModalParams` with the desired modal parameters. See [here](https://github.com/determined-ai/determined/pull/6188/files#diff-d09d5b7b58ed7d9b0ac4ddfe33e47cd302eace6e2ff9dbcfe3ae60b4b731300d) for an example

## Design Checklist [figma](https://www.figma.com/file/Pqs6TVjjHQohXSwJ9qaqwH/Determined-UI-Kit?node-id=261-8511&t=tEs7CQlABlWrS1k7-0)

feel free to add to this
- [ ] correct background/foreground color for danger icon on danger modal
- [ ] icons wrong size?
- [ ] padding between the different regions correct?
- [x] submitting state for button
- [x] header links (e.g. for relevant entity)
- [x] footer links (e.g. for relevant documentation)
- [x] click on background to close
- [x] "danger modal"
- [x] providing modals for icons
- [x] sizes

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: webui/react/src/modal/useProjectEditModal.tsx
Comment (ClaireNeveu) [inline webui/react/src/modal/useProjectEditModal.tsx:72]: Why do you need to duplicate the form state into the observable? If your goal is just to call a submit with the form data could you instead pass the form into the hook or have the form register itself using context?

File: webui/react/src/modal/useModal.tsx
Comment (ClaireNeveu) [inline webui/react/src/modal/useModal.tsx:23]: Can we make this small/medium/large? Don't see a reason to abbreviate here.

File: webui/react/src/modal/useProjectEditModal.tsx
Comment (ClaireNeveu) [reply webui/react/src/modal/useProjectEditModal.tsx:72]: I think if you pass in the form instance you can use [`form.getFieldsError()`](https://ant.design/components/form#forminstance) to check if fields have errors and disable the submit based on that.

**Review by ClaireNeveu — CHANGES_REQUESTED** [review]
I noticed two big design issues off the bat:
1. When the modal is submitting a form contained within should have its fields disabled
2. The modals are not responsive and go outside the window on smaller screens

File: webui/react/src/pages/DesignKit.tsx
Comment (ClaireNeveu) [inline webui/react/src/pages/DesignKit.tsx:1786]: Doesn't `JSX.IntrinsicAttributes` include things like `style` and `className`?

File: webui/react/src/modal/useModality.tsx
Comment (trentwatt) [inline webui/react/src/modal/useModality.tsx:153]: RE: 
> When the modal is submitting a form contained within should have its fields disabled

@johnkim-det something like replacing this with `<Spinner spinning={isSubmitting}>{modalBody}</Spinner>` could work.

File: webui/react/src/modal/useModality.tsx
Comment (trentwatt) [reply webui/react/src/modal/useModality.tsx:78]: The current pattern in the case of a submission failure is to close the modal and show a toast notification (the modal component itself should have already done all the validation it can).

The logic to show the toasts is currently in the various handle submit functions. As it, this would maintain that.

Comment (ClaireNeveu) [reply webui/react/src/modal/useModality.tsx:78]: I recognize this is the current design pattern but it's not really a good one. I'd like to open the possibility of handling the error without closing the modal.

Comment (trentwatt) [reply webui/react/src/modal/useModality.tsx:78]: in that have do you have a preference as to the API? 

one option would be to have the `handleSubmit` not catch their errors and then have a `handleSubmitError (e: Error) => void`, although this would add another quirk to the way the creators of specific modal components have to do things.

Comment (ClaireNeveu) [reply webui/react/src/modal/useModality.tsx]: Generally my preference is to expose callbacks but if we wanted to keep this promise-based approach we could just handle a failed promise.

**Comment by johnkim-det** [comment]
Updated with:
- Simplified component API
- Error handling for submit handler
- Fix for modal width issue
- Submission state for modal body
- Misc. refactoring

**Review by ClaireNeveu — APPROVED** [review]
Design looks good to me. Please wait for engineering review from @EmilyBonar before merging.

File: webui/react/src/components/kit/Modal.tsx
Comment (ClaireNeveu) [reply webui/react/src/components/kit/Modal.tsx:126]: Oh, this is a good callout. Can you add a required input to the modal in the UI Kit?

Comment (ClaireNeveu) [reply webui/react/src/components/kit/Modal.tsx:126]: One design note: there should be a tooltip on the disabled submit button explaining why it is disabled. Ideally this would aggregate the validation errors in the form but I know AntD's form API isn't super robust so if you can't do that can it can say "Address validation errors before proceeding".

File: webui/react/src/components/kit/Button.module.scss
Comment (EmilyBonar) [inline webui/react/src/components/kit/Button.module.scss:2]: question: What is this doing?

File: webui/react/src/components/kit/Button.tsx
Comment (EmilyBonar) [inline webui/react/src/components/kit/Button.tsx:29]: Look into using `ConditionalWrapper` for this.

File: webui/react/src/components/kit/Form.module.scss
Comment (EmilyBonar) [inline webui/react/src/components/kit/Form.module.scss:1]: super nit: I think we typically use camelCase instead of PascalCase for our classnames (https://khalilstemmler.com/blogs/camel-case-snake-case-pascal-case/)

File: webui/react/src/components/kit/Button.module.scss
Comment (johnkim-det) [reply webui/react/src/components/kit/Button.module.scss:2]: antd applies a margin to the submit button, which doesn't get applied when there's a wrapping element around the button, so the intent here was to make sure that spacing is preserved. but looking at it again after you asked about it, I think this is actually done by the antd styles for the modal, so i'll move it there. thanks!

**Comment by johnkim-det** [comment]
updated with styling fixes.

while setting color for the dangerous button, also noticed that there are a lot of Button variations that we may not need. went ahead and added them all to the DesignKit page, and created WEB-1079 to revisit them.

**Review by ClaireNeveu — APPROVED** [review]
Looks great! Thank you @johnkim-det and @trentwatt for the hard work!

**Pull request merged**
